src/gc-alloc-profiler.cpp

```diff
@@ -0,0 +1,139 @@
+// This file is a part of Julia. License is MIT: https://julialang.org/license
+
+#include "gc-alloc-profiler.h"
+
+#include "julia_internal.h"
+#include "gc.h"
+
+#include <string>
+#include <vector>
+
+using std::string;
+using std::vector;
+
+struct jl_raw_backtrace_t {
+    jl_bt_element_t *data;
+    size_t size;
+};
+
+struct jl_raw_alloc_t {
+    jl_datatype_t *type_address;
+    jl_raw_backtrace_t backtrace;
+    size_t size;
+};
+
+// == These structs define the global singleton profile buffer that will be used by
+// callbacks to store profile results. ==
+struct jl_per_thread_alloc_profile_t {
+    vector<jl_raw_alloc_t> allocs;
+};
+
+struct jl_alloc_profile_t {
+    double sample_rate;
+
+    vector<jl_per_thread_alloc_profile_t> per_thread_profiles;
+};
+
+struct jl_combined_results {
+    vector<jl_raw_alloc_t> combined_allocs;
+};
+
+// == Global variables manipulated by callbacks ==
+
+jl_alloc_profile_t g_alloc_profile;
+int g_alloc_profile_enabled = false;
+jl_combined_results g_combined_results; // Will live forever.
+
+// === stack stuff ===
+
+jl_raw_backtrace_t get_raw_backtrace() {
+    // A single large buffer to record backtraces onto
+    static jl_bt_element_t static_bt_data[JL_MAX_BT_SIZE];
```

Comment:
This is not thread safe. You should be (re)using the existing buffer in the ptls to be safe here.

Reply:
Whoops. Yeah, we need a buffer per thread…

Did you mean to (re)use this buffer? https://github.com/vilterp/julia/blob/c12aca890a8ae387d11db0b54351f8b61305c00b/src/julia_threads.h#L247

It seems like we should add our own (here or as a global) to avoid colliding with that… We've shied away from adding stuff to ptls to avoid breaking stuff or adding overhead, but maybe it's better than globals?

Funny our multithreaded test didn't catch this; maybe the stack traces were just garbage.

Reply:
As long as you avoid re-using this buffer in the interval between an error filling this buffer, and then allocating the exception stack, I think you should be okay here.

Reply:
The backtrace would be essentially the same, just a bit longer perhaps

Reply:
Thank you @vtjnash for pointing this out! 😅 haha it feels pretty silly that we missed that. Thanks for the help!!

The suggestion to use a buffer in ptls makes sense. I agree with @vilterp that sharing that buffer makes me slightly nervous, but reading through your description that sounds totally fine 👍 cool, thank you!

Reply:
Profiling using a different buffer entirely

Reply:
perfect, thanks.

Reply:
One more question: is it okay to write over this buffer, since it seems like it's used to scan for roots?:
https://github.com/JuliaLang/julia/blob/7ccc83ebaf0966c5782fe1d52a431ba51cc35472/src/task.c#L344-L345

https://github.com/JuliaLang/julia/blob/7ccc83ebaf0966c5782fe1d52a431ba51cc35472/src/task.c#L618-L623

I think that it's fine, from what i can read, but just want to double check one more time.

Also, i've opened a PR for this, here: https://github.com/JuliaLang/julia/pull/44114